### PR TITLE
Model config watcher support.

### PR DESCRIPTION
**Pull request by tlm**
Adds modelconfig watcher support.
    
    With this change we are introducing the concept of watching for model
    config changes via a watcher to the modelconfig.Service.
    
    This is needed by several api facades moving to dqlite.
    
    - Introduced a testing namespace watcher that can be fed with data
      changes via a channel. This allows services to be tested with out mock
      state. It is a very basic implementation at the moment to satisfy it's
      one use. I suspect this this will get more work and changes over time.
    - Moved the service testing state into the services package for model
      config. This is the testing mock states is making a state that
      conforms to the interface describe to in the service package and not
      that of the state package.
    - Updated the model config services to start testing the watcher. This
      package is still missing several tests but I will wait for API
      integration commits to include them to confirm the stability of the
      established contracts.
    - Adds change log triggers for model config table in the model database.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

Just run associated unit tests for the moment.

## Documentation changes

N/A

## Links

**Jira card:** JUJU-4853

### CODE REVIEW

File: domain/modelconfig/service/service_test.go [hunk 1]
```diff
@@ -43,8 +43,15 @@ func (s *serviceSuite) TestSetModelConfig(c *gc.C) {
...
-	st := statetesting.NewState()
-	svc := NewService(defaults, st)
+	st := testing.NewState()
+	defer st.Close()
+
+	svc := NewService(defaults, st, st)
+
+	watcher, err := svc.Watch()
+	c.Assert(err, jc.ErrorIsNil)
+	changes := <-watcher.Changes()
```
Comment (wallyworld) [inline domain/modelconfig/service/service_test.go:53]: We should never use naked selects

File: domain/modelconfig/service/service_test.go [hunk 2]
```diff
@@ -60,4 +67,9 @@ func (s *serviceSuite) TestSetModelConfig(c *gc.C) {
 		"secret-backend": "auto",
 		"logging-config": "<root>=INFO",
 	})
+
+	changes = <-watcher.Changes()
```
Comment (wallyworld) [inline domain/modelconfig/service/service_test.go:71]: ditto

File: domain/servicefactory/model.go [hunk 3]
```diff
@@ -18,12 +19,13 @@ type ModelFactory struct {
 // Config returns the model's configuration service. A ModelDefaultsProvider
 // needs to be supplied for the model config service. The provider can be
 // obtained from the controller service factory model defaults service.
-func (f *ModelFactory) Config(
+func (s *ModelFactory) Config(
```
Comment (wallyworld) [inline domain/servicefactory/model.go:22]: Why did the receiver name change?

File: domain/testing/watcher.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+// noop.
+func (f *NamespaceWatcherFactory) FeedChange(
+	namespace string,
+	_ changestream.ChangeType,
+	change []string) {
+	ch, exists := f.WatcherChans[namespace]
+	if !exists {
+		return
+	}
+	ch <- change
```
Comment (wallyworld) [inline domain/testing/watcher.go:43]: no naked channel operations

File: domain/testing/watcher.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+	if !exists {
+		f.WatcherChans[namespace] = make(chan []string, 1)
+		ch = f.WatcherChans[namespace]
+	}
+	if f.InitialStateFunc != nil {
+		state, err := f.InitialStateFunc(initialStateQuery)
+		if err != nil {
+			return nil, fmt.Errorf("fetching initial state for watcher: %w", err)
+		}
+		ch <- state
```
Comment (wallyworld) [inline domain/testing/watcher.go:62]: ditto

**Review by wallyworld — COMMENTED** [review]
This is introducing a lot of new testing infrastructure for state and watchers when we already have a core watchertest package with has channel driven mocks for strings and notify watchers. These existing mocks can be used in the service layer tests. It did land previously, but the in memory state mock also seems extravagant given the established pattern elsewhere (eg the externalcontrollers reference implementation) is to use gomocks
```
//go:generate go run go.uber.org/mock/mockgen -package service -destination state_mock_test.go github.com/juju/juju/domain/externalcontroller/service State

func TestPackage(t *testing.T) {
	gc.TestingT(t)
}
```
It bother me a lot we're starting to take divergent approaches to testing depending on who is writing the code. Juju has suffered greatly from this over the years. It would be suboptimal IMHO to repeat mistakes of the past.

Also the naked channel operations need to be fixed. We expect these to be wrapped like so
```
select {
    case <-ch.Changes():
    case <-time.After(jujutesting.LongWait):
	c.Fail()
}
```

File: domain/modelconfig/service/service_test.go (hunk 1 above)
Comment (tlm) [reply domain/modelconfig/service/service_test.go:53]: Agreed.

File: domain/servicefactory/model.go (hunk 3 above)
Comment (tlm) [reply domain/servicefactory/model.go:22]: Was just trying to align with the controller one.

File: domain/testing/watcher.go (hunk 4 above)
Comment (tlm) [reply domain/testing/watcher.go:43]: I've plumbed up the context like I should have done :)

File: domain/testing/watcher.go (hunk 5 above)
Comment (tlm) [reply domain/testing/watcher.go:62]: Plumbed up context.

**Comment by tlm** [comment]
Thanks @wallyworld for the review,

As for the testing approach we have agreed to role this way for the moment and discuss more next week about how we want to do this going forward.

I will update this PR should a different testing strategy be required.

File: domain/modelconfig/service/service_test.go [hunk 6]
```diff
@@ -60,4 +74,13 @@ func (s *serviceSuite) TestSetModelConfig(c *gc.C) {
 		"secret-backend": "auto",
 		"logging-config": "<root>=INFO",
 	})
+
+	select {
+	case changes = <-watcher.Changes():
+	case <-ctx.Done():
+		c.Fatal(ctx.Err())
+	}
```
Comment (wallyworld) [inline domain/modelconfig/service/service_test.go:82]: Interesting new approach, let's discuss this too to see how it stacks up.

**Review by wallyworld — APPROVED** [review]

**Pull request merged**